Replace curl usage with requests

zane-techempower opened the pull request:
_Early PR for a chunk of in-progress work_

These changes are centered around addressing #1568 which aims to use the native, and excellent Python [requests](http://docs.python-requests.org/en/latest/) package which prevents the need of using `curl` through Python. Notably this will improve Windows compatibility by removing the curl requirement. These changes touch the framework test types and their super class `framework_test_type` (which previously had the `_curl` and `_curl_response_body` methods). I have taken the liberty to improve the formatting of the files that I touch to be more closely in line with PEP8 in the interest of a cleaner TFB toolset.

- May fix #1600, I am checking that url's are correctly formed, and I am working in that area of files
- Fixes #1625 
- Will help address #1511 by breaking verification functionality into smaller and easier-to-test chunks. I have been keeping unit testing in mind as I've been working through this.
- Addresses #1534 by giving better warning messages. I have addressed concerns about [frameworks including charset in `Content-Type` field](https://github.com/TechEmpower/FrameworkBenchmarks/issues/1534#issuecomment-95766652) by adding an explicit check for the case. This makes the toolset `WARN`-city for certain frameworks that always append the charset.

- [x] Replace `_curl` and `_curl_response_body` with a similar method that utilizes the `requests` module
- [x] Go through test type files and:
    - Change their header checks for `requests`
    - Change their body checks for `requests`
    - Refactor checking logic to be more informative (mostly for `WARN`'s) and helpful to developers trying to add a framework
    - PEP 8 them, 4 space indentation, formatting, break into smaller functions
- [x] Refactor repeated/shared code
- [x] Refactor verification functions to be more grouped together
- [x] All test types inherit from `FrameworkTestType` directly (previously queries and updates were inheriting from `DBTestType` in order to share validation functions)
- [ ] Implement unit tests for the verification of headers and response bodies

This is Windows Compatibility / Round 12 work

hamiltont commented:
!!! Very cool !!!

My only nitpick - why don't we use `verifyBody` instead of `verifyObject` for the method name - it seems more precise to me as the return value is not always a JSON object e.g. plaintext

BTW, I was thinking that we could add a few "common" verify* methods to framework_test_type to avoid code duplication, such as methods like `verify_response_is_json`, `verify_json_has_top_level_keys([key1,key2])`, etc. Thoughts?

zane-techempower commented:
> `verifyBody` instead of `verifyObject`

Sounds good to me

> BTW, I was thinking that we could add a few "common" verify* methods to framework_test_type to avoid code duplication, such as methods like `verify_response_is_json`, `verify_json_has_top_level_keys([key1,key2])`, etc. Thoughts?

Currently stuff isn't very DRY, I've just been working through the changes on each file. I will do a nice refactor job at the end to pull out the common verify methods like you mentioned. This will also make unit testing as easy as possible. Good suggestions and certainly where I want to go with this.

zane-techempower commented:
Starting tomorrow I will not be able to commit time to this task, or other tfb tasks due to being assigned a full-time project. Aside from lack of unit tests, I am happy with the current state of this PR. I have cleaned up the `test_types` directory and moved much of it's functionality to `verifications.py` where it is more detailed (better warning and failure messages) and easier to understand (code formatting/shorter functions/DRYer).

Hopefully I will have time to at least start/set up the testing suite for these changes before round 12. I will accept PR's though!
